crates/sui-core/src/consensus_adapter.rs

```diff
@@ -895,6 +903,40 @@ impl ConsensusAdapter {
             .with_label_values(&[tx_type])
             .inc();
     }
+
+    /// Waits for transactions to appear either to consensus output or been executed via a checkpoint (state sync).
+    async fn await_consensus_or_checkpoint(
+        self: &Arc<Self>,
+        transaction_keys: Vec<SequencedConsensusTransactionKey>,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) {
+        let notifications = FuturesUnordered::new();
+        for transaction_key in transaction_keys {
+            let transaction_digests = if let SequencedConsensusTransactionKey::External(
+                ConsensusTransactionKey::Certificate(digest),
+            ) = transaction_key
+            {
+                vec![digest]
+            } else {
+                vec![]
+            };
+
+            notifications.push(async move {
```

Comment:
optional: why not aggregate all transaction keys and digests, wait on all of them, and avoid the extra FuturesUnordered? Are there advantages with only waiting on individual transactions?

Reply:
Good question: that was a slight optimisation to make sure that we have the full flexibility to ack the transactions from the independent paths (consensus & checkpoints). If we do the aggregation - and that was my initial implementation - then we'll basically mark them as "processed" whenever all of them have been processed either via checkpoint or via consensus (in bulk). I know that it might be quite a detail here, but I decided to go with this implementation for full flexibility.